File: Difficulty: Easy/Two Swaps/two-swaps.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
public:
    bool checkSorted(const vector<int> &arr) {
        vector<int> outOfOrd;  // Stores elements that are not in order
        int cnt = 0;  // Count of elements that are out of order
        
        // Identify elements that are out of order
        for (int i = 0; i < arr.size(); ++i) {
            if (arr[i] != i + 1) {
                outOfOrd.push_back(arr[i]);
                ++cnt;
            }
        }
        
        // If no elements are out of order, the array is already sorted
        if (cnt == 0) return true;
        // If the number of out-of-order elements is not 3 or 4, return false
        else if (cnt != 4 && cnt != 3) return false;
        
        vector<int> ref;  // Reference vector to compare the out-of-order elements
        for (const int val : outOfOrd) ref.push_back(val);
        
        int n = cnt;  // Number of elements out of order
        sort(ref.begin(), ref.end());  // Sort the out-of-order elements
        
        // Check if the sorted out-of-order elements match the original sequence
        for (int i = 0; i < n; ++i) {
            if (ref == outOfOrd && cnt == 4) return i % 2 == 0;  // For 4 out-of-order elements, check the alternating pattern
            if (ref == outOfOrd && cnt == 3) return true;  // If 3 elements are out of order, return true
            
            int tmp = ref[0];
            // Rotate the reference vector to check for another possible matching order
            for (int j = 0; j < n - 1; ++j) ref[j] = ref[j + 1];
            ref[n - 1] = tmp;
        }
        
        // Sort the reference vector in descending order and repeat the check
        sort(ref.begin(), ref.end(), greater<int>());
        for (int i = 0; i < n; ++i) {
            if (ref == outOfOrd && cnt == 4) return i % 2 == 0;  // For 4 out-of-order elements, check the alternating pattern
            int tmp = ref[0];
            // Rotate the reference vector to check for another possible matching order
            for (int j = 0; j < n - 1; ++j) ref[j] = ref[j + 1];
            ref[n - 1] = tmp;
        }
        
        // Return false if no valid configuration is found
        return false;
    }
};


//{ Driver Code Starts.
int main() {
    string ts;
    getline(cin, ts);
    int t = stoi(ts);
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        bool ans = obj.checkSorted(arr);
        if (ans) {

            cout << "true" << endl;
        } else {
            cout << "false" << endl;
        }

        cout << "~" << endl;
    }
}

// } Driver Code Ends

File: Difficulty: Easy/Two Swaps/two-swaps.js
//{ Driver Code Starts
process.stdin.resume();
process.stdin.setEncoding('utf-8');
let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', () => {
    inputString = inputString.trim().split('\n');
    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    const t = parseInt(readLine(), 10);
    const solution = new Solution();

    for (let i = 0; i < t; i++) {
        const arr = readLine().split(' ').map(Number); // Read array input
        const result = solution.checkSorted(arr);
        console.log(result ? "true" : "false");
        console.log("~");
    }
}

// } Driver Code Ends

class Solution {
    // Function to swap once in case of mismatch
    doOneSwap(n, arr) {
        // Iterate through the array to find the first mismatch
        for (let i = 0; i < n; i++) {
            // If an element is not in its correct position
            if (arr[i] !== i + 1) {
                // Search for the element that should be in the current position
                for (let j = i + 1; j < n; j++) {
                    // When the element is found, perform a swap
                    if (arr[j] === i + 1) {
                        [arr[i], arr[j]] = [arr[j], arr[i]]; // Swap elements
                        return; // Exit after one swap
                    }
                }
            }
        }
    }

    // Function to check if sorted or can be sorted with a maximum of two swaps
    checkSorted(arr) {
        let n = arr.length;
        let misMatch = 0;

        // Count the number of mismatches between the current element and its correct position
        for (let i = 0; i < n; i++) {
            if (arr[i] !== i + 1) {
                misMatch++;
            }
        }

        // Edge case: if the array has only one element, it cannot be sorted
        if (n === 1) {
            return false;
        }

        // If the array is already sorted, or if there are exactly 3 mismatches (which can be fixed with 2 swaps)
        if (misMatch === 0 || misMatch === 3) {
            return true;
        }

        // If there are not exactly 4 mismatches, it is not possible to sort with two swaps
        if (misMatch !== 4) {
            return false;
        }

        // Perform two swaps to check if sorting is possible
        this.doOneSwap(n, arr);
        this.doOneSwap(n, arr);

        // Final check to ensure the array is sorted
        for (let i = 0; i < n; i++) {
            if (arr[i] !== i + 1) {
                return false; // If still not sorted, return false
            }
        }

        return true; // Return true if the array is sorted
    }
}
